### ISSUE TITLE
Saving Sliced Tensors writes entire unsliced Tensor

### ISSUE THREAD

**Issue by khdlr** [pytorch/pytorch#40157]
## 🐛 Bug

Calling `torch.save` on a sliced tensor seems to write the entire unsliced tensor to disk.

## To Reproduce

```
import torch

array = torch.zeros(1000000)
slice = array[:1000]
clone = slice.clone()

torch.save(array, 'array.pt') # 3.9MB on disk
torch.save(slice, 'slice.pt') # 3.9MB on disk
torch.save(clone, 'clone.pt') # 4.3KB on disk
```

## Expected behavior

`torch.save` only writes the bytes to disk that are part of the slice. The file-sizes of `slice.pt` and `clone.pt` in the above example should be equal.

## Environment
```
PyTorch version: 1.5.0
Is debug build: No
CUDA used to build PyTorch: 10.2

OS: Arch Linux
GCC version: (GCC) 10.1.0
CMake version: version 3.17.3

Python version: 3.8
Is CUDA available: Yes
CUDA runtime version: 10.2.89
GPU models and configuration: GPU 0: GeForce RTX 2060 SUPER
Nvidia driver version: 440.82
cuDNN version: /usr/lib/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip3] numpy==1.18.5
[pip3] torch==1.5.0
[pip3] torchsummary==1.5.1
[conda] Could not collect
```

cc @jlin27

**Comment by ailzhang** [comment]
This is actually expected behavior - We save the whole storage instead of tensor itself. We do so to preserve storage sharing in serialization. This allows tensors which share storage before serialization still share storage after loading from checkpoint.
```
>>> import torch
>>>
>>> array = torch.zeros(1000000)
>>> slice = array[:1000]
>>> torch.save([slice, array[:10]], 'share.pt')
>>> a, b = torch.load('share.pt')
>>> a[0]
tensor(0.)
>>> b[0]
tensor(0.)
>>> b[0] = 1
>>> a[0]
tensor(1.)
```
But if you clone before saving, it creates new storage so the two tensors will no longer share storage. 

Hope this explains, please feel free to reopen if you have followup questions.

**Comment by khdlr** [comment]
I see the usecase of saving tensors that share storage, which is a great feature imho.

Is this documented anywhere? I feel the current behaviour is counterintuitive, especially to new users who may not know how tensors map to memory regions.

Also in the above example, the advanced behaviour isn't needed and leads to disk space being wasted. Is it possible to automatically detect whether writing the entire underlying storage to disk is needed?

**Comment by khdlr** [comment]
Let me share my usecase, maybe it will make clear why I'm so confused by this issue.

I work on segmentation of satellite images. In order to allow smooth, shuffled data loading, I load the big satellite image (~8000x8000 pixels) into torch once, and write out smaller tiles to disk in order to load them later in random order for training.
So I wrote the script to do this tiling and let it run. But then after some time, my hard disk was full because every single tile was 1GB in size, when it should have been only a couple of MB according to my calculations.

Of course now I know how to fix this issue, but it still caught me by surprise and led to some tedious debugging on why the script was doing this.

**Comment by ailzhang** [comment]
@khdlr Nice catch about documentation this - I'll send a patch soon for it since I've only seen answers in issues but not in our official documentation. 
W.r.t automatic detection and save less in the case of no storage sharing, my first guess is it's possible to implement it with some additional checks. Adding this issue for triage review to see whether we want this feature.

**Comment by albanD** [comment]
I think the detection would be a useful feature to have but it will be BC-breaking (I can see weird use case were it will actually change the behavior slightly).

**Comment by ezyang** [comment]
We discussed this and we're going just go for a doc change.
